tgg/annotations/get_ensembl_db_info.py

```diff
@@ -22,53 +24,168 @@
 homo_sapiens_variation_103_38
 """
 
-def get_canonical_transcripts(database=CURRENT_ENSEMBL_DATABASE):
-    gene_id_to_canonical_transcript_id = {}
-    with pymysql.connect(host="useastdb.ensembl.org", user="anonymous", database=database) as conn:
-        with conn.cursor() as c:
-            columns = ["gene.stable_id", "transcript.stable_id"]
-            columns_str = ", ".join(columns)
-            c.execute(f"SELECT {columns_str} FROM gene LEFT JOIN transcript ON canonical_transcript_id = transcript_id")
 
-            for row in c:
-                d = dict(zip(columns, row))
-                gene_id = d['gene.stable_id']
-                transcript_id = d['transcript.stable_id']
-                if gene_id in gene_id_to_canonical_transcript_id:
-                    other_id = gene_id_to_canonical_transcript_id[gene_id]
-                    raise Exception(f"{gene_id} has more than 1 canonical transcript: {transcript_id}, {other_id}")
-                gene_id_to_canonical_transcript_id[gene_id] = transcript_id
+def get_gene_id_to_transcript_metadata(
+        database=CURRENT_ENSEMBL_DATABASE,
+        only_protein_coding=False,
+        only_canonical_transcripts=False):
+    """Retrieves a dictionary containing gene_id => a list of dictionaries each of which
+    contains information about one transcript that belongs to that gene.
 
-    return gene_id_to_canonical_transcript_id
+    Args:
+        database (str): The Ensembl database name (eg. "homo_sapiens_core_107_38")
+        only_protein_coding (bool): If True, only return protein-coding genes and protein-coding transcripts
+        only_canonical_transcripts (bool): If True, only return canonical transcripts
```

Comment:
I often note what the defaults are for the args, not sure if that would be helpful here

Reply:
For me this adds redundant info that needs to be maintained or risks becoming out-of-date if defaults change and I forget to update the docs later. If we decide to publish auto-generated html docs for these functions, the default values should be embedded in the docs automatically. If we continue looking at these docs directly in source code, the default values are easy to see as is.